### ISSUE TITLE
Questions regarding fuzzing workflow using AFL++ / Best practices?

### ISSUE THREAD

**Issue by fuzzah** [AFLplusplus/AFLplusplus#258]
Sup. My questions as of today:
1. Do we have some place for silly questions like this one?
2. Do we have some 'best-practice' fuzzing workflows using new AFL++ features? / How do maintainers use AFL++ to maximize fuzzing efficiency? (there are so many options and new features though tutorials on the net are just some 'stories of success' literally only demonstrating some fuzzing capabilities using ANY viable techniques, like nobody really cares for THE BEST fuzzing workflow)
 
**My proposal**: the idea on the most bug-findy fuzzing workflow with AFL++ when target sources available which includes 4 stages (target builds). Please prove me wrong, I'll be happy to learn some better techniques. My wheel-reinventing(?) method mostly applies to huge targets. For small applications you may just go with all the 'heavy' options AFL++ offers.
 
**So, "teh process"**
 
Stage 1. Check if it worth fuzzing at all.
1. Ensure you're testing the most recent version of your target application.
2. If your target has some tests, build the target with ASAN & UBSAN and run tests.
3. If your target's code is utterly broken and every single run with sanitizers provides you lots of different crash messages that get you terrified af **AND** if you're by yourself interested in fixing this code, you can start off fuzzing with AFL_HARDEN **and** libdislocator instead of ASAN & UBSAN for this and further stages to start fixing at least something. If you're fuzzing someone else's code and not interested in fixing, then just go ahead with another "fuzzing police" report bombing poor repo's issues with huge AddressSanitizer messages.
 
Stage 2. Discover paths.
1. Correctly implement persistent `__AFL_LOOP` reading data from stdin and putting it to tested code. This should happen as close to tested code as possible.
2. Use afl-clang-fast with whitelist. **If** target fails to compile with clang, use afl-gcc-fast with whitelist. **If** compilation still **somehow** fails, use afl-gcc. Be sure to use llvm/clang/gcc versions not less than 9.
3. Use LLVM_LAF environment variables to improve coverage. If something fails to compile, try removing these variables one by one.
4. At first, use only basic AFL++ instrumentation (no sanitizers, no AFL_HARDEN), for now this is only to quickly discover paths.
5. Use tmpfs for faster .cur_input rewriting and reading _(also doable with tools like anything-sync-daemon when the whole fuzzing directory is put onto a ramdisk and then synchronized with physical drive from time to time)_
6. Run AFL++ fuzzers in parallel with both master and slave instances.
7. Use power schedules (`-p fast` is good enough).
8. Use dictionaries. To generate afl-compatible dictionary one may use some **grep**ping of source code.
9. Use a couple of real test cases as inputs (don't just go for tutorial classics like `echo 1234 > in/1`).
10. Minimize inputs with cmin and tmin.
11. Don't fully rely on Radamsa as it will reduce your total execs/second. If you still want to use it, start separate slave instance with exclusive Radamsa mutator (-RR).
12. Stop fuzzing when there are enough paths discovered. You may do so after 1 hour or you may do so after a couple of weeks, mostly it depends on amount of size of your target's code base.
 
Stage 3. Hunt all the bugs.
1. Recompile your target using ASAN & UBSAN (no LLVM_LAF needed this time).
2. Run sanitizer-compiled targets on discovered paths using testcases residing in `queue` sub-directories to identify if anything crashes. Easiest is to just use commands like `find ... | xargs ...` .
 
Stage 4. Check code coverage you've got so far.
1. Recompile your target with code coverage collection. No AFL++ instrumentation needed this time.
2. Run your coverage-collecting target on discovered paths in the same manner as in stage 3.
3. Generate and check your code coverage reports.
4. Bad coverage? Make better dictionary and find better initial test cases, then start over with Stage2.
 
Is this the most effective strategy? I want some critique suggestions for improvements.

**Comment by chibinz** [comment]
+1
AFL++ is iterating very fast and many features are added constantly. 
- Some features are inherited from AFL and may not work so well nowadays(deprecate them?).
- Other features work well under specific circumstances but not all the time(There are readmes for each individual feature, and the --help print out. Maybe gather them together, better docs?)
- There are also features that suit most fuzzing scenarios. (Make them default, as suggested by issue #185)
And most importantly how do we combine and organize this features to get the best out of AFL++? Maintainers probably use AFL++ a lot and know what works and what doesn't. Sharing your fuzzing strategy or workflow definitely helps:)

**Comment by vanhauser-thc** [comment]
Some input from me:

Stage 1 stept 2 + 3 ... I would not waste my time on it, but it might be a useful move for horribly bad targets. but on such target you dont need fuzzing as the project is already broken.

Stage 2 step 2: Try afl-clang-lto/afl-clang-lto++ first (it is the fastest instrumentation and has 0 map collision/pollution), if it this does not build then use afl-clang-fast/afl-clang-fast++ with AFL_LLVM_INSTRIM=1 (nstrim gives better performance and less map pollution compared to normal llvm_mode) 
step 6: should clarify to always be 1 master only, not more not less
step 8: libtokencap
step 11: totally agree
insert step 12: use redqueen/cmplog (see llvm_mode/README.cmplog.md)

Original step 12 + stage 3 - I would say no to this. I would rather recommend the following:
for every afl-fuzz instance compile its own binary (unless llvm_mode LTO is used. normal+instrim+afl_gcc+afl-gcc-fast instrument with random ID, so compiling the target multiple times enhance the chance for path discovery due avoiding map collisions) , and a 1/3 or 1/4 should be ASAN+UBSAN, the rest not.
And it makes sense to have some instances using afl-smart, some FairFuzz, one radamsa only ... and if a slave does not work - replace it.

**Comment by vanhauser-thc** [comment]
We should wait for more comments from @hexcoder- @andreafioraldi @domenukk and maybe other users/reader, and then you can create a docs/README.best_practices.md and send a PR :)

**Comment by sijohans** [comment]
First when i started using AFL i had something like this:
```c
int main(void)
{
    uint8_t buffer[8192];
    ssize_t bytes_read = read(0, buffer, sizeof(buffer));

    if (bytes_read <= 0)
    {
        return 0;
    }

    parse_my_data(buffer, bytes_read);
    return 0;
}
```
I'm just showing a very basic example here. Assume that the **prase_my_data** is buggy and reads perhaps 5-10 more bytes than the input size parameter. Assume further that **bytes_read < sizeof(buffer) - 10)**. Here, AFL would not detect a crash, even if using ASAN. To improve this, i copy the input data to a newly allocated buffer. E.g.:
```c
int main(void)
{
    uint8_t buffer[8192];
    ssize_t bytes_read = read(0, buffer, sizeof(buffer));

    if (bytes_read <= 0)
    {
        return 0;
    }

    uint8_t *cpy = malloc(bytes_read);
    if (NULL == cpy)
    {
        return 0;
    }
    memcpy(cpy, buffer, bytes_read):
    parse_my_data(cpy, bytes_read);
    free(cpy);
    return 0;
}
```
I'm not sure if i have seen this tip somewhere. What are you take on this approach?

Another thing I've also been using is the return code of the main program. If the input is considered as valid, it would return 0, otherwise non zero value. Using this i have been using the return codes on the fuzz output to compare how other implementations (e.g. in other languages) considers the input. E.g.:
```c

extern bool parse_my_data(const uint8_t *input, size_t size);

int main(void)
{
    uint8_t buffer[8192];
    ssize_t bytes_read = read(0, buffer, sizeof(buffer));

    if (bytes_read <= 0)
    {
        return 0;
    }

    uint8_t *cpy = malloc(bytes_read);
    if (NULL == cpy)
    {
        return 0;
    }
    memcpy(cpy, buffer, bytes_read):
    bool valid_input = parse_my_data(cpy, bytes_read);
    free(cpy);
    return (valid_input) ? 0 : -1;
}
```

I've found these two approaches very useful. Are you doing something similar?
Other things that would be nice:
* Have a provided main entry that provides the above functionality. I used libFuzzer **LLVMFuzzerTestOneInput** as idea and made a similar approach with AFL.
* Have some interactive script that allows one to easy to different fuzz operations on one or multiple targets. E.g.:
  * Switch mode (ASAN, HARDEN), libFuzzer, ...?
  * cmin, tmin
  * dry run on all found paths and generate coverage report
  * copy found paths to input
  * allows for parallel mode

I have a very nasty shell script that does this for a project. I'm using CMake for bulding and just creating different build directories and changing compiler/flags when invoking the CMake script. It looks something like:
```
Fuzzing tool

    [1] Select target [fuzz_my_target]
    [2] Select mode [build_libfuzzer_asan/test/fuzz]
    [3] Select num cores [4]
    [4] Start fuzz (Press CTRL+C to break)
    [5] cmin and tmin
    [6] Copy output to input
    [7] Reset
    [8] Dry run with coverage
    [0] Quit

    Select: 4
```
That script is as mentioned, very nasty. Would be nice to split it into different pieces to that it could be easy to use in e.g. CI.

One this i also realize is that you continuous learn new stuff. And the things you did yesterday you want to do different the next day.

**Comment by hexcoder-** [comment]
Nice roundup!

One thing that was a bit irritating to me at first was the realization that many of AFL++'s reported many 'unique crashes' are very often manifestations of the same bug triggered on multiple ways.
So triaging and evaluating the crashing test cases can be a lot of duplicated (manual?) work.

I work mostly with open source, so I went into the habit of regarding a crashing test case only relevant when the crashing code location was different from anything I had triaged before.
I did this because in 99% the bug location was very near to the crash location.

**Comment by domenukk** [comment]
I wonder if `-D_FORTIFY_SOURCE=2` would spot most of the bug in the first example 🤔

**Comment by domenukk** [comment]
I wonder if `-D_FORTIFY_SOURCE=2` could spot bugs similar to the first example of @sijohans 🤔
If the function isn't too complex, at least?

**Comment by floyd-fuh** [comment]
+1. It is hard to keep track at all.

Suggestion so this is not too much documentation overhead: Examples with a list of recommended features to check out. Maybe just explain strategies that can be derived from https://www.fuzzbench.com/reports/2020-04-08-aflplusplus-and-2020-04-01/index.html and use those targets as the examples.

Off topic: I'm super excited about the new features and really looking forward to multi-threading in afl-fuzz. A command line switch that starts multiple threads with different strategies and switches strategies on its own sounds doable.

**Comment by vanhauser-thc** [comment]
I agree too much documentation is an issue.
I already started on a picture that shows the important features and decisions to make, basically an SVG that shows the command line / environment option to set when using that feature

off topic: compile llvm-11 and use afl-clang-lto and AFL_LLVM_LTO_AUTODICTIONARY .... :)

**Comment by domenukk** [comment]
No activity here in a while. Closing :)
